### PR TITLE
Add BackOffHandler to FailedRecordTracker

### PR DESCRIPTION
**Pull request by Marincek**
Added new BackOffHandler interface. The idea is to decouple the FailedRecordTracker with the behavior on next back off time. 
Currently the FailedRecordTracker is sleeping the thread, and there is now way to change that behaviour.

Related issue : https://github.com/spring-projects/spring-kafka/issues/2262

### CODE REVIEW

**Review by selalerercapitolis — APPROVED** [review]

**Review by artembilan — CHANGES_REQUESTED** [review]
I wonder if existing `KafkaConsumerBackoffManager` abstraction can work here as well, without introducing anything else...

@tomazfernandes, WDYT ?

**Comment by garyrussell** [comment]
@artembilan That is for pausing a single partition, in this case we need to pause the entire container.

@Marincek Are you planning on providing the implementation that pauses and resumes the container?

**Comment by tomazfernandes** [comment]
> I wonder if existing `KafkaConsumerBackoffManager` abstraction can work here as well, without introducing anything else...
> 
> @tomazfernandes, WDYT ?

Missed this notification, sorry @artembilan. As @garyrussell pointed out, the `KafkaConsumerBackOffManager` is meant to back off a single partition, but we probably could just retrieve all assignments for that container and loop through them. 

One important thing is we need to make sure a `idlePartitionEventInterval` is set for the container - maybe we could look into providing a default value for that.

Now that you brought this up, and considering this `pause until` pattern appears here and there, I think in the future we might look into leveraging the `BackOffManager` to handle all such logic - the `nack` feature comes to mind as well for example - so that we don't need to implement it over and over again and have a centralized place to handle all partitions. Maybe we could auto configure a `BackOffManager` bean for that and inject where needed.

But I defer to @garyrussell what would be the proper solution for this particular case.

**Comment by garyrussell** [comment]
@Marincek use `./gradlew check` to build the project (including running checkstyle).

**Comment by garyrussell** [comment]
Sorry for the delay; I have squashed and merged your commits to this branch https://github.com/garyrussell/spring-kafka/tree/GH-2262

with this polishing commit: https://github.com/garyrussell/spring-kafka/commit/17f8f85d8c532268fa17ba506bfd23be388f5389

I will merge both after my commit is reviewed by @artembilan

**Comment by garyrussell** [comment]
Merged and cherry-picked; thanks for the contribution!

**Pull request closed without merging**
